src/Launcher.vala

```diff
@@ -28,6 +31,56 @@ public class Dock.Launcher : Gtk.Button {
         windows = new GLib.List<AppWindow> ();
         get_style_context ().add_provider (css_provider, Gtk.STYLE_PROVIDER_PRIORITY_APPLICATION);
 
+        var pinned_menu_item = new MenuItem (null, null);
+        pinned_menu_item.set_attribute_value ("custom", "pinned-item");
+
+        var close_menu_item = new MenuItem (null, null);
+        close_menu_item.set_attribute_value ("custom", "close-item");
+
+        var action_section = new Menu ();
+        foreach (var action in app_info.list_actions ()) {
+            action_section.append (
+                app_info.get_action_name (action),
+                MainWindow.ACTION_PREFIX + MainWindow.LAUNCHER_ACTION_TEMPLATE.printf (app_info.get_id (), action)
+            );
+        }
+
+        var model = new Menu ();
+        model.append_item (pinned_menu_item);
+        model.append_item (close_menu_item); // Placeholder, currently doesn't work
+        if (action_section.get_n_items () > 0) {
+            model.append_section (null, action_section);
+        }
+
+        var pinned_label = new Gtk.Label ("Keep in Dock") {
+            xalign = 0,
+            hexpand = true
+        };
+
+        var pinned_check_button = new Gtk.CheckButton ();
+
+        var pinned_box = new Gtk.Box (HORIZONTAL, 3);
+        pinned_box.append (pinned_label);
+        pinned_box.append (pinned_check_button);
+
+        var pinned_button = new Gtk.ToggleButton () {
+            child = pinned_box
+        };
+        pinned_button.add_css_class (Granite.STYLE_CLASS_MENUITEM);
+
+        close_button = new Gtk.Button () {
+            visible = false,
+            child = new Gtk.Label (_("Close")) { halign = START }
+        };
+        close_button.add_css_class (Granite.STYLE_CLASS_MENUITEM);
+
+        popover = new Gtk.PopoverMenu.from_model (model) {
+            position = TOP
+        };
```

Comment:
This will close the menu when it loses focus on Wayland

```suggestion
        popover = new Gtk.PopoverMenu.from_model (model) {
            autohide = true,
            position = TOP
        };
```